src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockImporter.scala

```diff
@@ -0,0 +1,350 @@
+    val currentBest = getStartingBlockNumber()
+    log.debug(
+      "Picking blocks, current best block with hash is {}, best by blockchain is {}",
+      currentBest,
+      blockchain.getBestBlockNumber())
+
+    val msg = state.resolvingBranchFrom.fold[BlockFetcher.FetchMsg](
+      BlockFetcher.PickBlocks(blocksBatchSize)
+    )(
+      from => BlockFetcher.StrictPickBlocks(from, currentBest)
+    )
+
+    fetcher ! msg
+  }
+
+  private def importBlocks(blocks: NonEmptyList[Block]): ImportFn =
+    importWith({
+      log.debug("Attempting to import blocks starting from {}", Block.number(blocks.head))
+      Future
+        .successful(resolveBranch(blocks))
+        .flatMap({
+          case Right(blocksToImport) => handleBlocksImport(blocksToImport)
+          case Left(resolvingFrom) => Future.successful(ResolvingBranch(resolvingFrom))
+        })
+    })
+
+  private def handleBlocksImport(blocks: List[Block]): Future[NewBehavior] =
+    tryImportBlocks(blocks)
+      .map(value => {
+        val (importedBlocks, errorOpt) = value
+        log.debug("imported blocks {}", importedBlocks.map(Block.number).mkString(","))
+
+        errorOpt match {
+          case None =>
+            self ! PickBlocks
+            Running
+          case Some(err) =>
+            log.error(
+              "block import error {}, current best block is {} and best block with hash is {}",
+              err,
+              blockchain.getBestBlockNumber(),
+              getStartingBlockNumber())
+            val notImportedBlocks = blocks.drop(importedBlocks.size)
+
+            err match {
+              case e: MissingNodeException =>
+                fetcher ! BlockFetcher.FetchStateNode(e.hash)
+                ResolvingMissingNode(NonEmptyList(notImportedBlocks.head, notImportedBlocks.tail))
+              case _ =>
+                val invalidBlockNr = Block.number(notImportedBlocks.head)
+                fetcher ! BlockFetcher.InvalidateBlocksFrom(invalidBlockNr, err.toString)
+                self ! PickBlocks
+                Running
+            }
+        }
+      })
+
+  private def tryImportBlocks(blocks: List[Block], importedBlocks: List[Block] = Nil)(
+      implicit ec: ExecutionContext): Future[(List[Block], Option[Any])] =
+    if (blocks.isEmpty) {
+      Future.successful((importedBlocks, None))
+    } else {
+      val restOfBlocks = blocks.tail
+      ledger
+        .importBlock(blocks.head)
+        .flatMap {
+          case BlockImportedToTop(_) =>
+            tryImportBlocks(restOfBlocks, blocks.head :: importedBlocks)
+
+          case ChainReorganised(_, newBranch, _) =>
+            tryImportBlocks(restOfBlocks, newBranch.reverse ::: importedBlocks)
+
+          case DuplicateBlock | BlockEnqueued =>
+            tryImportBlocks(restOfBlocks, importedBlocks)
+
+          case err @ (UnknownParent | BlockImportFailed(_)) =>
+            log.debug(s"block ${Block.number(blocks.head)} import failed")
```

Comment:
start the message with capital letter and maybe consider extracting `blocks.head` to some val